### ISSUE TITLE
Issue when using WildFly OpenSSL 1.1.3.Final with Jetty Server 9.4.34

### ISSUE THREAD

**Issue by jalberti** [eclipse/jetty.project#5650]
**Jetty version**

WildFly OpenSSL: 1.1.3.Final
Jetty Server: 9.4.34.v20201102

**Java version**

Java 8

**Question**
Cross posting here, see also https://github.com/wildfly-security/wildfly-openssl/issues/89, I was wondering if the jetty community has seen this issue before somewhere, and maybe somebody has an idea what the root cause could be?

The server is configured to use OpenSSL.
In the client, when using the default JSSE provider, everything works as expected, and the certificates are available in the request in the server when reading the attribute javax.servlet.request.X509Certificate.

The issue is present when I set the client to use OpenSSL. When using OpenSSL, the request is completed and the response is received by the client, but the server shows two strange outcomes:

the certificates are not available in the request object; AND

an NPE results from an invalid session:

```
2020-11-05 16:55:44.993:WARN:oejs.SecureRequestCustomizer:qtp1440047379-21: Unable to get secure details 
java.lang.NullPointerException: while trying to get the length of a null array loaded from local variable 'b'
	at org.eclipse.jetty.util.TypeUtil.toHexString(TypeUtil.java:524)
	at org.eclipse.jetty.server.SecureRequestCustomizer$SslAttributes.getSslSessionData(SecureRequestCustomizer.java:386)
	at org.eclipse.jetty.server.SecureRequestCustomizer$SslAttributes.<init>(SecureRequestCustomizer.java:336)
	at org.eclipse.jetty.server.SecureRequestCustomizer.customize(SecureRequestCustomizer.java:283)
	at org.eclipse.jetty.server.SecureRequestCustomizer.customize(SecureRequestCustomizer.java:210)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:379)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:556)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:540)
	at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:395)
	at org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:161)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:773)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:905)
	at java.lang.Thread.run(Thread.java:836)

====== CERTIFICATES: null
```

Reproduction code is attached in the wildfly github issue.

**Comment by gregw** [comment]
At the very least we should avoid the NPE.  But the question is, what should we do with a request on an SSL connection that does not have a full session?  Reject the request? Allow the request to continue, but not marked as secure?    Or do we just say this is some invalid behaviour in the SSL layer and failing with an NPE is as reasonable as anything else.

**Comment by gregw** [comment]
Also whilst fixing this, lets fix the repeated calls to `getSslSessionData()` [here](https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java#L336-L340)

**Comment by sbordet** [comment]
I don't think we should be too defensive when passed non-valid `SSLSession` objects from a TLS implementation -- we trust the TLS implementation to do things right, otherwise we would need to write a million tons of defensive code.

**Comment by jalberti** [comment]
Looks like the wildfly colleagues have now added a test case to assert their behavior for themselves, and their test case is passing, any chance somebody is able to help to point out why this is not working inside of Jetty ... https://github.com/wildfly-security/wildfly-openssl/pull/90/files?

See also https://issues.redhat.com/browse/WFSSL-60

**Comment by johannes-altiscale** [comment]
wildfly is unable to reproduce the issue, the issue is suspected to be inside Jetty's usage/configuration, the issue has been migrated to https://issues.redhat.com/browse/WFSSL-65.

_"With the reproducer with Jetty, OpenSSLSessionContext#sessionCreatedCallback is not being invoked when WildFly OpenSSL is used on the client side. Normally, this is invoked when a new session has been negotiated and session caching is enabled. This will need further investigation to pinpoint the reason for this, it could be related to Jetty itself and/or the way the Jetty and Apache SSL contexts are being configured in the reproducer."_

**Comment by jalberti** [comment]
@joakime and @sbordet thank you for looking into this issue, we made multiple passes at this, but we have not made much progress, looking at the latest release from the WildFly colleagues, we still see the same issue with Jetty. Would you be able to help pin point the root cause? Thank you! And btw. a Happy New Year to everybody that reads this, and who follows the Gregorian calendar. :)

**Comment by gregw** [comment]
I'm not even sure where to look to address this one.

Perhaps an instrumented version of jetty can be written that records the exact sequence of calls made to SslEngine API to accept the connection.     I test server could them be written to perform exactly this sequence to see if the issue can be reproduced without Jetty involved.

@jalberti is this something you can do or would you need our assistance?

**Comment by sbordet** [comment]
@jalberti I looked at the implementations of Conscrypt, OpenJDK, Netty, etc. and all return a non-null `byte[]`.

Either you are using an old implementation that could return `null`, or otherwise I don't see how that bug is possible (modulo JIT/GC bugs).

This is not something that depends on Jetty: we get the session id from standard APIs, and that should not be `null`.
I would double check exactly what SSL implementation you're using, and see if the vendor/version you're using could return `null`.

**Comment by gregw** [comment]
@sbordet is there any chance we get confused and think the SslEngine has completed its handshake before it really has?

**Comment by sbordet** [comment]
@gregw even so, which I doubt, all implementations I have seen return an empty, non-null, `byte[]` if the session id is not available.

**Comment by jalberti** [comment]
We made some progress with this issue, we still do not understand why it is working or not, but we can confidently state that when we have a client that uses in the client hello message the extension `server_ticket` and the server actually in response generates a "new session ticket" ... the reported issue can be observed. If the client does not request, or the server does not generate.. the handshake works fine within the jetty context and client certs are available on the server side. 

E.g.
- Conscrypt 2.5.1 on the client with Wildfly 2.1.2.Final on the server ... works 
- Wildfly on the client and Conscrypt on the server ... works
- Making changes in Wildfly to disable `server_ticket` support ... option `SSL.SSL_OP_NO_TICKET` ... works

When looking at the packet capture we can see a successful handshake, but on the server side we see the reported exception. We still do not see how to proceed further here. We will obviously also relay this information to the Wildfly colleagues. Thanks for all your help!

For reference ... https://github.com/wildfly-security/wildfly-openssl/blob/2b2311c98c135d443e891a2b0e77a9b579ea017e/java/src/main/java/org/wildfly/openssl/OpenSSLContextSPI.java#L169-L181

At the moment in our local configuration ... `OpenSSL Version OpenSSL 1.1.1h  22 Sep 2020` ... `disableSessionTicketsSupported` always comes back as false, but if explicitly disabled despite being reported as not supported in the first place, everything works fine.

**Comment by jalberti** [comment]
The issue is fixed with Wildfly 2.1.3.Final -> https://github.com/wildfly-security/wildfly-openssl/pull/98
